File: sample_code_data/mergesort/java/22.java
package com.example;

import java.lang.annotation.Target;
import java.util.Arrays;
import java.util.Scanner;

public class Main {

    private static int[] merge(int[] left, int[] right) {
        int i = 0, j = 0;
        int[] merged = new int[left.length+right.length];
        int k = 0;
        while (i < left.length && j < right.length) {
            if (left[i] <= right[j])
                merged[k] = left[i++];
            else
                merged[k] = right[j++];
            k++;
        }

        if (i < left.length) {
            while (i < left.length) {
                merged[k++] = left[i++];
            }
        }else {
            while (j < right.length) {
                merged[k++] = right[j++];
            }
        }

        return merged;
    }

    private static int[] mergeSort(int[] ar) {
        if (ar.length <= 1)
            return ar;

        // Split the array into two halves
        int mid = ar.length / 2;
        int[] left = Arrays.copyOfRange(ar, 0 , mid);
        int[] right = Arrays.copyOfRange(ar, mid, ar.length);
        // Sort the first half
        int[] sortedLeft = mergeSort(left);
        // Sort the second half
        int[] sortedRight = mergeSort(right);
        // Merge
        int[] merged = merge(sortedLeft, sortedRight);
        return merged;
    }

    public static void main(String[] args) {
        int[] ar = {18, 6, 9, 1, 4, 15, 12, 5, 6, 7, 11};
        int[] result = mergeSort(ar);
        printArray(result);
    }

    private static void printArray(int[] ar) {
        for(int n: ar){
            System.out.print(n+" ");
        }
        System.out.println("");
    }
}

File: code/mergesort/cpp/22.cpp
#include <bits/stdc++.h>

using namespace std;

void merge(int A[],int s,int e){
    int p = s,k = 0;
    int arr[e - s + 1];
    int m = (s+e)/2;
    int j = m+1;
    for(int i=s;i<=e;i++){
        if(j > e){
            arr[k++] = A[p++];
        }
        else if(p > m){
            arr[k++] = A[j++];
        }
        else if(A[p] < A[j]){
            arr[k++] = A[p++];
        }else{
            arr[k++] = A[j++];
        }
    }
    for(int i=0;i<k;i++){
        A[s++] = arr[i];
    }
}

void mergesort(int arr[],int s,int e){
    if(s < e){
        int mid = (s+e)/2;
        mergesort(arr,s,mid);
        mergesort(arr,mid+1,e);
        merge(arr,s,e);
    }
}

int main(){
    int arr[] = {64,25,12,22,11};
    int n = sizeof(arr)/sizeof(int);
    mergesort(arr,0,n-1);
    for(int i=0;i<n;i++){
        cout<<arr[i]<<" ";
    }
return 0;
}
